test/org/centos/contra/Infra/Utils.groovy

```diff
@@ -0,0 +1,347 @@
+
+        when:
+        infraUtils.generateTopology(instance, 1, "somedir")
+
+        then:
+        noExceptionThrown()
+        1 * getPipelineMock("writeFile")([file: "PinFile", text: """1_openstack:
+  topology:
+    topology_name: "os-single-new"
+    resource_groups:
+      - resource_group_name: "os-server-new"
+        resource_group_type: "openstack"
+        resource_definitions:
+          - name: "$instance_name"
+            role: "os_server"
+            flavor: "m1.small"
+            image: "CentOS-7-x86_64-GenericCloud-1612"
+            count: 1
+            keypair: ci-factory
+            fip_pool: 10.8.240.0
+            networks:
+              - atomic-e2e-jenkins-test
+            
+        credentials:
+          filename: "openstack.creds"
+          profile: ci-rhos
+"""])
+    }
+
+    def "key files can be generated based on provider type"() {
+        when:
+        infraUtils.createKeyFile('aws')
+        infraUtils.createKeyFile('beaker')
+        infraUtils.createKeyFile('openstack')
+        then:
+        1 * getPipelineMock("file.call").call(['credentialsId': "aws.creds", 'variable': 'CREDENTIAL_FILE_NAME'])
+        1 * getPipelineMock("file.call").call(['credentialsId': "beaker.creds", 'variable': 'CREDENTIAL_FILE_NAME'])
+        1 * getPipelineMock("file.call").call(['credentialsId': "openstack.creds", 'variable': 'CREDENTIAL_FILE_NAME'])
+
+    }
+
+    def "ssh keys can be generated for a given provider"() {
+        when:
+        infraUtils.createSSHKeyFile('aws', 'someContainer')
+        infraUtils.createSSHKeyFile('beaker', 'someContainer')
+        infraUtils.createSSHKeyFile('openstack', 'someContainer')
+        then:
+        1 * getPipelineMock("sshUserPrivateKey.call").call(['credentialsId': 'aws.ssh', 'keyFileVariable': 'SSH_FILE_NAME', 'passphraseVariable': 'SSH_PASSPHRASE', 'usernameVariable': 'SSH_USERNAME'])
+        1 * getPipelineMock("sshUserPrivateKey.call").call(['credentialsId': 'beaker.ssh', 'keyFileVariable': 'SSH_FILE_NAME', 'passphraseVariable': 'SSH_PASSPHRASE', 'usernameVariable': 'SSH_USERNAME'])
+        1 * getPipelineMock("sshUserPrivateKey.call").call(['credentialsId': 'openstack.ssh', 'keyFileVariable': 'SSH_FILE_NAME', 'passphraseVariable': 'SSH_PASSPHRASE', 'usernameVariable': 'SSH_USERNAME'])
+    }
+
+    def "context files generated by linchpin can be parsed"() {
+
+        getPipelineMock("readJSON")(_) >> {
+            return ['validJSON': true]
+        }
+        when:
+        def result = infraUtils.parseDistilledContext('someFile')
+        then:
+        result == ['validJSON': true]
+    }
+
+    def "inventory file can be generated"() {
+        def instance = infraUtils.createAwsInstances(config.infra.provision.cloud.aws as HashMap)[0]
+        when:
+        infraUtils.generateInventory(instance, [:])
+        then:
+        1 * getPipelineMock("writeFile").call(['file': 'Mock Generator for [WORKSPACE]/inventory', 'text': '\n\n'])
+
+    }
+}
```

Comment:
My personal preference is having a newline at the end of the file. But that can be discussed on how others feel about it

Reply:
I'll say that newline or no newline is less concerning for me, whatever is chosen, just ensure it's consistent throughout the test files.